OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs

```diff
@@ -0,0 +1,506 @@
+			public readonly long CabDescriptorOffset;
+			public readonly uint CabDescriptorSize;
+
+			public CommonHeader(Stream reader)
+			{
+				Version = reader.ReadUInt32();
+				VolumeInfo = reader.ReadUInt32();
+				CabDescriptorOffset = (long)reader.ReadUInt32();
+				CabDescriptorSize = reader.ReadUInt32();
+			}
+		}
+
+		struct CabDescriptor
+		{
+			public readonly long FileTableOffset;
+			public readonly uint FileTableSize;
+			public readonly uint FileTableSize2;
+			public readonly uint DirectoryCount;
+
+			public readonly uint FileCount;
+			public readonly long FileTableOffset2;
+
+			public CabDescriptor(Stream reader, CommonHeader commonHeader)
+			{
+				reader.Seek(commonHeader.CabDescriptorOffset + 12, SeekOrigin.Begin);
+				FileTableOffset = (long)reader.ReadUInt32();
+				/*    unknown  */ reader.ReadUInt32();
+				FileTableSize = reader.ReadUInt32();
+
+				FileTableSize2 = reader.ReadUInt32();
+				DirectoryCount = reader.ReadUInt32();
+				/*   unknown  */ reader.ReadBytes(8);
+				FileCount = reader.ReadUInt32();
+
+				FileTableOffset2 = (long)reader.ReadUInt32();
+			}
+		}
+
+		struct FileDescriptor
+		{
+			public readonly ushort	Flags;
+			public readonly uint	ExpandedSize;
+			public readonly uint	CompressedSize;
+			public readonly uint	DataOffset;
+
+			public readonly byte[]	MD5;
+			public readonly uint	NameOffset;
+			public readonly ushort	DirectoryIndex;
+			public readonly uint	LinkToPrevious;
+
+			public readonly uint	LinkToNext;
+			public readonly byte	LinkFlags;
+			public readonly ushort	Volume;
+			public readonly string	Filename;
+
+			public FileDescriptor(Stream reader, long tableOffset)
+			{
+				Flags = reader.ReadUInt16();
+				ExpandedSize = reader.ReadUInt32();
+				/*    unknown   */ reader.ReadUInt32();
+				CompressedSize = reader.ReadUInt32();
+
+				/*  unknown */ reader.ReadUInt32();
+				DataOffset   = reader.ReadUInt32();
+				/*  unknown */ reader.ReadUInt32();
+				MD5          = reader.ReadBytes(16);
+
+				/*   unknown  */ reader.ReadBytes(16);
+				NameOffset     = reader.ReadUInt32();
+				DirectoryIndex = reader.ReadUInt16();
+				/*   unknown  */ reader.ReadBytes(12);
+				LinkToPrevious = reader.ReadUInt32();
+				LinkToNext = reader.ReadUInt32();
+
+				LinkFlags = reader.ReadBytes(1)[0];
+				Volume = reader.ReadUInt16();
+				var posSave = reader.Position;
+
+				reader.Seek(tableOffset + NameOffset, SeekOrigin.Begin);
+				Filename = reader.ReadASCIIZ();
+				reader.Seek(posSave, SeekOrigin.Begin);
+			}
+		}
+
+		class CabReader : IDisposable
+		{
+			readonly FileDescriptor fileDes;
+			public uint RemainingArchiveStream;
+			public uint RemainingFileStream;
+			readonly uint index;
+			readonly string commonName;
+			ushort volumeNumber;
+			Stream cabFile;
+
+			public CabReader(FileDescriptor fileDes, uint index, string commonName)
+			{
+				this.fileDes = fileDes;
+				this.index = index;
+				this.commonName = commonName;
+				volumeNumber = fileDes.Volume;
```

Comment:
Minor style nit: It would be clearer to define this as `volumeNumber = fileDes.Volume - 1;`